feat: adding endpoints for files and uploads

vladimirivic opened the pull request:
Summary:
Adds spec definitions for file uploads operations.

This API focuses around two high level operations:
* Initiating and managing upload session
* Accessing uploaded file information

Usage examples:

To start a file upload session:
```
curl -X POST https://localhost:8321/v1/files \
-d '{
   "key": "image123.jpg',
   "bucket": "images",
   "mime_type": "image/jpg",
   "size": 12345
}'

# Returns
{
  “id”: <session_id>
  “url”: “https://localhost:8321/v1/files/session:<session_id>”,
  "offset": 0,
  "size": 12345
}

```

To upload file content to an existing session
```
curl -i -X POST "https://localhost:8321/v1/files/session:<session_id> \
  --data-binary @<path_to_local_file>

# Returns
{
  "key": "image123.jpg",
  "bucket": "images",
  "mime_type": "image/jpg",
  "bytes": 12345,
  "created_at": 1737492240
}

# Implementing on server side (Flask example for simplicity):
@app.route('/uploads/{upload_id}', methods=['POST'])
def upload_content_to_session(upload_id):
    try:
        # Get the binary file data from the request body
        file_data = request.data

        # Save the file to disk
        save_path = f"./uploads/{upload_id}"
        with open(save_path, 'wb') as f:
            f.write(file_data)
        return {__uploaded_file_json__}, 200
    except Exception as e:
        return 500

```

To read information about an existing upload session
```
curl -i -X GET "https://localhost:8321/v1/files/session:<session_id>

# Returns
{
  “id”: <session_id>
  “url”: “https://localhost:8321/v1/files/session:<session_id>”,
  "offset": 1024,
  "size": 12345
}
```

To list buckets
```
GET /files

# Returns
{
  "data": [
     {"name": "bucket1"},
     {"name": "bucket2"},
   ]
}
```

To list all files in a bucket
```
GET /files/{bucket}

# Returns
{
  "data": [
    {
      "key": "shiba.jpg",
      "bucket": "dogs",
      "mime_type": "image/jpg",
      "bytes": 82334,
      "created_at": 1737492240,
    },
    {
      "key": "persian_cat.jpg",
      "mime_type": "image/jpg",
      "bucket": "cats",
      "bytes": 39924,
      "created_at": 1727493440,
    },
  ]
}
```

To get specific file info
```
GET /files/{bucket}/{key}

{
  "key": "shiba.jpg",
  "bucket": "dogs",
  "mime_type": "image/jpg",
  "bytes": 82334,
  "created_at": 1737492240,
}

```

To delete specific file
```
DELETE /files/{bucket}/{key}

{
  "key": "shiba.jpg",
  "bucket": "dogs",
  "mime_type": "image/jpg",
  "bytes": 82334,
  "created_at": 1737492240,
}

```

vladimirivic commented on llama_stack/apis/uploads/uploads.py:
```diff
@@ -0,0 +1,37 @@
+        mime_type: str,
+    ) -> Optional[UploadResponse]: ...
+
+    @webmethod(route="/uploads/{upload_id}", method="POST")
+    async def upload_content(
+        self,
+        upload_id: str,
+    ) -> Optional[FileResponse]: ...
```
This one should have request body generated as: 

```
requestBody:
  content:
    application/octet-stream:
      schema:
        type: string
        format: binary
```

vladimirivic replied:
Solved the problem in generator by looking at the request body params and checking if the type is bytes to mark content type as octet-stream.

vladimirivic commented:
Status: currently discussing merging /uploads and /files

dineshyv commented on docs/openapi_generator/pyopenapi/generator.py:
```diff
@@ -545,15 +547,35 @@ def _build_operation(self, op: EndpointOperation) -> Operation:
+            # use the application/octet-stream media type.
+            if (
+                len(op.request_params) == 1
+                and len(op.request_params[0]) == 2
```
should this be specific to only 1 param?

vladimirivic replied:
Yes, only 1 param when recv upload as application/octet-stream

vladimirivic replied:
This was a hack. I found a better solution, look for `raw_bytes_request_body`

yanxi0830 commented on llama_stack/apis/files/files.py:
```diff
@@ -0,0 +1,51 @@
+@trace_protocol
+class Files(Protocol):
+    @webmethod(route="/files/{bucket}", method="GET")
+    async def list_bucket(
```
Could we add some documentation on what each param is used for?

yanxi0830 commented on docs/openapi_generator/pyopenapi/generator.py:
```diff
@@ -545,15 +547,35 @@ def _build_operation(self, op: EndpointOperation) -> Operation:
-                description=doc_params.get(request_name),
-                required=True,
-            )
+            # If there is only one request parameter and it is of bytes type
+            # use the application/octet-stream media type.
```
n00b qs: how does `application/octet-stream` work, it always accepts byte type with 1 param?

yanxi0830 replied:
https://github.com/openai/openai-openapi/blob/acea67a5045a6873c6eb2573a6be0e0a7b092ec6/openapi.yaml#L10104-L10133 -- how does `application/octet-stream` work differently compared with `application/json`

vladimirivic replied:
This means that the request body will be raw bytes

booxter commented on llama_stack/apis/files/files.py:
```diff
@@ -0,0 +1,72 @@
+
+@runtime_checkable
+@trace_protocol
+class Files(Protocol):
```
A terminology question: what makes it a file (which is usually part of a file system)? Isn't it something more generic, like a blob, or an artifact, or an object (like in s3)?

vladimirivic replied:
Looking from users' perspective, when writing code and doing some file processing, they will be picking their local files to upload to Llama Stack server so they will be also files there.

booxter replied:
Context of the question is that I would like to have an API to store artifacts of job execution (logs, outputs), so I was thinking about adding an API for this. It seems like this `files` API proposal is trying to achieve the same thing. But in my mind, these are just "artifacts". Yes, they may later be "downloaded" onto a user's file system, at which point they become "files" (with permissions, paths, metadata specific to the file system). But while it's "in the cloud" (on the llama-stack server), I think about these entities as something abstract (blobs, objects).

Of course, this is just my preference and others may have other thoughts. Just sharing the context here.

ashwinb replied:
@booxter yeah you are exactly right. the semantic of this API is indeed "blob storage" (aka S3) and so this would be used for artifacts as well. We could name this "/blobs" but there are some use cases (e.g., when dealing with inference chat completion or agents APIs where a user could be uploading files like PDFs also) where it is much more natural to call it "/files". As an example, OpenAI does it. 

For now, I think we should move ahead with this naming but realize it is indeed not perfect and somewhat confusing.

booxter commented on llama_stack/apis/files/files.py:
```diff
@@ -0,0 +1,72 @@
+    @webmethod(route="/files", method="POST")
+    async def create_upload_session(
+        self,
+        bucket: str,
```
What is a bucket? Is it a container for files (blobs)? Who defines these?

vladimirivic replied:
Bucket will be like a label under which we will have a unique key. It is simply a way to group keys under a common name. For example label can be: images, datasets, videos etc. 

Will update comments in the code.

vladimirivic replied:
You can look at buckets as single level namespaces where you add your keys. Bucket name should be unique so if you upload two files under and specify the same bucket, those files will be listed when you run GET /files/{bucket}.

ashwinb replied:
@jaideepr97 As @vladimirivic noted above, I think the same rough semantics as S3. If we need recursive retrievals, we could add an API for it or a parameter to the current retrieval API. Would that work? I don't think we want to create a full filesystem API here because it does not match with what underlying cloud providers typically provide.

booxter commented on llama_stack/apis/files/files.py:
```diff
@@ -0,0 +1,72 @@
+        self,
+        bucket: str,
+        key: str,
+        mime_type: str,
```
what are possible values for this field? should type names be enforced against a list of known types?

vladimirivic replied:
There is a long list of common types https://developer.mozilla.org/en-US/docs/Web/HTTP/MIME_types/Common_types but this can also be user defined

booxter reviewed (commented):
To be able to assess whether a proposed API makes sense, it helps to describe how this is going to be used. Is it possible to include some documentation and example in this PR? Thanks.

vladimirivic commented:
> Is it possible to include some documentation and example in this PR?

Certainly, will work on those and make updates soon. Thanks for the review!

vladimirivic commented on docs/_static/llama-stack-spec.yaml:
```diff
@@ -1581,6 +1673,35 @@ paths:
+            type: string
+      requestBody:
+        content:
+          application/octet-stream:
```
cc: Stainless Client SDK support

vladimirivic commented:
Extended webmethod decorator to allow describing when request body will be raw bytes. That way we can generate correct OpenAPI schema.

Removed the bytes param from upload_content_to_session. Updated doc strings.

booxter commented on llama_stack/apis/files/files.py:
```diff
@@ -0,0 +1,130 @@
+        """
+        ...
+
+    @webmethod(route="/files/{bucket}/{key:path}", method="DELETE")
```
The current scheme is:

```
POST /files/ - create file
POST /files/{upload_id} - upload file
GET /files/{bucket} - list files from the bucket
GET /files/{bucket}/{key:path} - get file
DELETE /files/{bucket}/{key:path} - delete file
```

I think this could be more symmetric:

```
POST /files/{bucket}/{key:path} - create file
POST /files/{bucket}/{key:path}/upload - upload file
GET /files/{bucket} - list files from the bucket
GET /files/{bucket}/{key:path} - get file
DELETE /files/{bucket}/{key:path} - delete file
```

(Note that now we pass key and bucket in URL when creating / uploading a file, the same way as we do when getting or deleting it.)

What do you think?

ashwinb replied:
@vladimirivic thoughts on this one?

vladimirivic replied:
Thanks for the proposal. One thing that makes it difficult to implement is that this route `POST /files/{bucket}/{key:path}/upload` is in conflict with `POST /files/{bucket}/{key:path}` since key can have the '/' char.

booxter commented on llama_stack/apis/files/files.py:
```diff
@@ -0,0 +1,130 @@
+    bucket: str
+    key: str
+    mime_type: str
+    url: str
```
Should this field be enforced as a proper URL? https://docs.pydantic.dev/2.2/usage/types/urls/

vladimirivic replied:
Will leave it as strong for now, as I am thinking of proposing internal url schema for accessing files internally, something like `lsfs:/{bucket}/{key}` (lsfs for Llama Stack file system)

booxter commented on llama_stack/apis/files/files.py:
```diff
@@ -0,0 +1,130 @@
+    """
+
+    bucket: str
+    key: str
```
Should any of these fields be constrained to be non-empty? https://docs.pydantic.dev/1.10/usage/types/#arguments-to-constr

vladimirivic replied:
I think we can, but I don't see any other examples in this repo. Maybe this is a good follow up to do across all APIs.

jaideepr97 reviewed (commented):
Thanks for the PR! 
I have a general question around how this API defintion extends to the remote case. If we needed to perform an action (like training or inferencing against a new model) on a machine that is separate from the one the server is running on, we would probably need to be able to retrieve checkpoints and such from that remote machine 

should something like that be handled in the implementation of a remote provider for this API, or should we treat this set of endpoints as geared to inline only, with a different set of endpoints for the remote case?

yanxi0830 reviewed (approved):
stamp to unblock
